corehq/apps/hqadmin/management/commands/load_domain_data.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import unicode_literals
+
+import gzip
+import json
+import os
+import warnings
+import zipfile
+from collections import defaultdict
+
+from django.apps import apps
+from django.conf import settings
+from django.core.management.base import BaseCommand, CommandError
+from django.core.management.color import no_style
+from django.core.serializers.python import (
+    Deserializer as PythonDeserializer,
+)
+from django.db import (
+    DatabaseError, IntegrityError, connections, router,
+    transaction,
+)
+from django.utils.encoding import force_text
+
+from corehq.form_processor.backends.sql.dbaccessors import ShardAccessor
+from corehq.sql_db.config import partition_config
+
+
+partitioned_model_id_fields = {
+    'form_processor.XFormInstanceSQL': 'form_id',
+    'form_processor.XFormAttachmentSQL': 'form',
+    'form_processor.XFormOperationSQL': 'form',
+}
+
+
+class Command(BaseCommand):
+    help = 'Loads data from the give file into the database.'
+    args = '<dump file path>'
+
+    def handle(self, dump_file_path, **options):
+        self.verbosity = options.get('verbosity')
+
+        # Keep a count of the installed objects
+        self.loaded_object_count = 0
+        self.total_object_count = 0
+
+        self.compression_formats = {
+            None: (open, 'rb'),
+            'gz': (gzip.GzipFile, 'rb'),
+            'zip': (SingleZipReader, 'r'),
+        }
+
+        if not os.path.isfile(dump_file_path):
+            raise CommandError("Dump file not found: {}".format(dump_file_path))
+
+        self.load_data(dump_file_path)
+
+        if self.verbosity >= 1:
+            if self.total_object_count == self.loaded_object_count:
+                self.stdout.write("Installed %d object(s)" % self.loaded_object_count)
+            else:
+                self.stdout.write("Installed %d object(s) (of %d)" %
+                                  (self.loaded_object_count, self.total_object_count))
+
+    def load_data(self, dump_file_path):
+        """
+        Loads data from a given file.
+        """
+        cmp_fmt = self.get_compression_format(os.path.basename(dump_file_path))
+        open_method, mode = self.compression_formats[cmp_fmt]
+        dump_file = open_method(dump_file_path, mode)
+        try:
+            if self.verbosity >= 2:
+                self.stdout.write("Installing data from %s." % dump_file_path)
+
+            objects = json.load(dump_file)
```

Comment:
I can't tell if this is possible from a quick look through documentation, but is it possible to write your own serializer for django? Something like jsonlines would allow you to go through line by line and not have to load a ginormous json structure at once. I think if its possible then the rest of your code could stay largely the same